src/charm.py

```diff
@@ -252,6 +333,312 @@ def _on_nginx_route_broken(self, _: Any) -> None:
+            return
+        relation = self._get_relation()
+        if relation is None:
+            self._cleanup()
+            self.unit.status = WaitingStatus("waiting for relation")
+            return
+        hostname = self._tls.get_hostname_from_csr(
+            tls_certificates_relation, self.app, event.certificate_signing_request
+        )
+        old_csr = self._tls.get_relation_data_field(
+            f"csr-{hostname}", tls_certificates_relation, self.app
+        )
+        private_key_dict = {}
+        if JujuVersion.from_environ().has_secrets:
+            secret = self.model.get_secret(label=f"private-key-{hostname}")
+            secret.grant(tls_certificates_relation)
+            private_key_dict["key"] = secret.get_content()["key"].encode()
+            private_key_dict["password"] = secret.get_content()["password"].encode()
+        else:
+            private_key_dict["key"] = self._tls.get_relation_data_field(
+                f"key-{hostname}", tls_certificates_relation, self.app
+            ).encode()
+            private_key_dict["password"] = self._tls.get_relation_data_field(
+                f"password-{hostname}", tls_certificates_relation, self.app
+            ).encode()
+        new_csr = generate_csr(
+            private_key=private_key_dict["key"],
+            private_key_password=private_key_dict["password"],
+            subject=hostname,
+        )
+        self.certificates.request_certificate_renewal(
+            old_certificate_signing_request=old_csr.encode(),
+            new_certificate_signing_request=new_csr,
+        )
+        self._tls.update_relation_data_fields(
+            {f"csr-{hostname}": new_csr.decode()}, tls_certificates_relation, self.app
+        )
+        peer_relation = self.model.get_relation("nginx-peers")
+        if peer_relation:
+            self._tls.update_relation_data_fields(
+                {f"csr-{hostname}": new_csr.decode()}, peer_relation, self.app
+            )
+
+    # This method is too complex but hard to simplify.
+    def _certificate_revoked(self, revoke_list: List[Union[str, None]]) -> None:  # noqa: C901
+        """Handle TLS Certificate revocation.
+
+        Args:
+            revoke_list: TLS Certificates list to revoke
+        """
+        tls_certificates_relation = self.model.get_relation("certificates")
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for peer relation to be created")
+            return
+        relation = self._get_relation()
+        if relation is None:
+            self._cleanup()
+            self.unit.status = WaitingStatus("waiting for relation")
+            return
+        for hostname in revoke_list:
+            old_csr = self._tls.get_relation_data_field(
+                f"csr-{hostname}", tls_certificates_relation, self.app
+            )
+            if not old_csr:
+                continue
+            if JujuVersion.from_environ().has_secrets:
```

Comment:
Shouldn't we managed the case when juju doesn't support secrets ?

Reply:
If we have secrets enabled, we currently store things in the secret and the tls relation databag, if we don't we just use the relation databag. So, the case is actually managed in this case. If what we want is to use only secrets and not write the relation databag when secrets are enabled, please let me know so I can make the changes.

Reply:
Well my idea is indeed to only store the secret ID in the relation databag when secrets are enabled, and fallback on peer relation otherwise. Concerning TLS relation that's a different story as the support should come from the library.
What do you think about this proposal ?

Reply:
Makes perfect sense. I'll implement it right away.